File: android/app/src/main/java/com/example/devicesync/crypto/SessionCrypto.kt
package com.example.devicesync.crypto

import android.util.Base64
import android.util.Log
import java.security.KeyFactory
import java.security.KeyPairGenerator
import java.security.interfaces.ECPublicKey
import java.security.spec.ECGenParameterSpec
import java.security.spec.X509EncodedKeySpec
import javax.crypto.Cipher
import javax.crypto.KeyAgreement
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * Manages ECDHE P-256 key exchange and AES-256-GCM encryption for session payloads.
 *
 * Lifecycle:
 * 1. Call [generateKeyPair] at session start to create the local ECDHE keypair.
 * 2. Send [getPublicKeyBase64] to the remote peer.
 * 3. When the peer's public key arrives, call [deriveSharedSecret] to compute the AES-256 key.
 * 4. Use [encrypt] / [decrypt] for all subsequent sensitive payloads.
 */
class SessionCrypto {

    companion object {
        private const val TAG = "SessionCrypto"
        private const val EC_ALGORITHM = "EC"
        private const val EC_CURVE = "secp256r1"   // P-256
        private const val KEY_AGREEMENT_ALGORITHM = "ECDH"
        private const val AES_ALGORITHM = "AES"
        private const val AES_GCM_TRANSFORMATION = "AES/GCM/NoPadding"
        private const val GCM_TAG_LENGTH_BITS = 128
        private const val GCM_IV_LENGTH_BYTES = 12
    }

    private var localKeyPair: java.security.KeyPair? = null
    private var sharedSecret: SecretKey? = null

    /** Whether a shared secret has been derived (key exchange is complete). */
    val isEstablished: Boolean
        get() = sharedSecret != null

    /**
     * Generate the local ECDHE P-256 key pair.
     * Call once when the WebSocket session is opened.
     */
    fun generateKeyPair() {
        val kpg = KeyPairGenerator.getInstance(EC_ALGORITHM)
        kpg.initialize(ECGenParameterSpec(EC_CURVE))
        localKeyPair = kpg.generateKeyPair()
        sharedSecret = null // reset in case of re-keying
        Log.d(TAG, "Generated ECDHE P-256 key pair")
    }

    /**
     * Returns the local public key as a Base64-encoded string (X.509 DER format).
     * This is what we send to the remote peer in the key_exchange message.
     */
    fun getPublicKeyBase64(): String {
        val pub = localKeyPair?.public
            ?: throw IllegalStateException("Key pair not generated. Call generateKeyPair() first.")
        return Base64.encodeToString(pub.encoded, Base64.NO_WRAP)
    }

    /**
     * Given the remote peer's Base64-encoded public key, derive the shared AES-256 secret.
     *
     * After this call, [isEstablished] returns true and [encrypt]/[decrypt] are usable.
     */
    fun deriveSharedSecret(remotePublicKeyBase64: String) {
        val remoteKeyBytes = Base64.decode(remotePublicKeyBase64, Base64.NO_WRAP)
        val keySpec = X509EncodedKeySpec(remoteKeyBytes)
        val keyFactory = KeyFactory.getInstance(EC_ALGORITHM)
        val remotePublicKey = keyFactory.generatePublic(keySpec) as ECPublicKey

        val keyAgreement = KeyAgreement.getInstance(KEY_AGREEMENT_ALGORITHM)
        keyAgreement.init(localKeyPair?.private
            ?: throw IllegalStateException("Key pair not generated."))
        keyAgreement.doPhase(remotePublicKey, true)

        // ECDH raw shared secret is 32 bytes for P-256; use directly as AES-256 key
        val rawSecret = keyAgreement.generateSecret()
        // Take first 32 bytes (SHA-256 of raw secret for uniform distribution)
        val digest = java.security.MessageDigest.getInstance("SHA-256")
        val derivedKey = digest.digest(rawSecret)
        sharedSecret = SecretKeySpec(derivedKey, AES_ALGORITHM)

        Log.d(TAG, "ECDHE shared secret derived successfully")
    }

    /**
     * Encrypt a plaintext string using AES-256-GCM.
     *
     * @return Base64-encoded ciphertext in format: `iv_base64:ciphertext_base64`
     */
    fun encrypt(plaintext: String): String {
        val key = sharedSecret
            ?: throw IllegalStateException("Shared secret not established. Complete key exchange first.")

        val cipher = Cipher.getInstance(AES_GCM_TRANSFORMATION)
        cipher.init(Cipher.ENCRYPT_MODE, key)
        val iv = cipher.iv // auto-generated 12-byte IV
        val ciphertext = cipher.doFinal(plaintext.toByteArray(Charsets.UTF_8))

        val ivBase64 = Base64.encodeToString(iv, Base64.NO_WRAP)
        val ctBase64 = Base64.encodeToString(ciphertext, Base64.NO_WRAP)
        return "$ivBase64:$ctBase64"
    }

    /**
     * Decrypt a ciphertext string produced by [encrypt].
     *
     * @param encryptedPayload format: `iv_base64:ciphertext_base64`
     * @return the original plaintext
     */
    fun decrypt(encryptedPayload: String): String {
        val key = sharedSecret
            ?: throw IllegalStateException("Shared secret not established. Complete key exchange first.")

        val parts = encryptedPayload.split(":", limit = 2)
        if (parts.size != 2) throw IllegalArgumentException("Invalid encrypted payload format")

        val iv = Base64.decode(parts[0], Base64.NO_WRAP)
        val ciphertext = Base64.decode(parts[1], Base64.NO_WRAP)

        val cipher = Cipher.getInstance(AES_GCM_TRANSFORMATION)
        val spec = GCMParameterSpec(GCM_TAG_LENGTH_BITS, iv)
        cipher.init(Cipher.DECRYPT_MODE, key, spec)
        val plainBytes = cipher.doFinal(ciphertext)

        return String(plainBytes, Charsets.UTF_8)
    }

    /**
     * Reset the crypto state (e.g. on disconnect).
     */
    fun reset() {
        localKeyPair = null
        sharedSecret = null
        Log.d(TAG, "Session crypto state reset")
    }
}

File: windows/WindowsAgent/SessionCrypto.cs
using System;
using System.Diagnostics;
using System.Security.Cryptography;
using System.Text;

namespace WindowsAgent
{
    /// <summary>
    /// Manages ECDHE P-256 key exchange and AES-256-GCM encryption for session payloads.
    ///
    /// Lifecycle:
    /// 1. Call <see cref="GenerateKeyPair"/> at session start.
    /// 2. Send <see cref="GetPublicKeyBase64"/> to the remote peer.
    /// 3. When the peer's public key arrives, call <see cref="DeriveSharedSecret"/>.
    /// 4. Use <see cref="Encrypt"/> / <see cref="Decrypt"/> for all sensitive payloads.
    /// </summary>
    public class SessionCrypto
    {
        private const int GcmTagLengthBytes = 16;   // 128 bits
        private const int GcmIvLengthBytes = 12;

        private ECDiffieHellman? _localKeyPair;
        private byte[]? _sharedSecret;

        /// <summary>
        /// Whether a shared secret has been derived (key exchange is complete).
        /// </summary>
        public bool IsEstablished => _sharedSecret != null;

        /// <summary>
        /// Generate the local ECDHE P-256 key pair. Call once per WebSocket session.
        /// </summary>
        public void GenerateKeyPair()
        {
            _localKeyPair?.Dispose();
            _localKeyPair = ECDiffieHellman.Create(ECCurve.NamedCurves.nistP256);
            _sharedSecret = null; // reset on re-key
            Debug.WriteLine("SessionCrypto: Generated ECDHE P-256 key pair");
        }

        /// <summary>
        /// Returns the local public key as a Base64-encoded string (X.509 SubjectPublicKeyInfo DER).
        /// This is what we send to the Android peer in the key_exchange message.
        /// </summary>
        public string GetPublicKeyBase64()
        {
            if (_localKeyPair == null)
                throw new InvalidOperationException("Key pair not generated. Call GenerateKeyPair() first.");

            var pubBytes = _localKeyPair.PublicKey.ExportSubjectPublicKeyInfo();
            return Convert.ToBase64String(pubBytes);
        }

        /// <summary>
        /// Given the remote peer's Base64-encoded public key (X.509 DER),
        /// derive the shared AES-256 secret.
        /// </summary>
        public void DeriveSharedSecret(string remotePublicKeyBase64)
        {
            if (_localKeyPair == null)
                throw new InvalidOperationException("Key pair not generated.");

            var remoteKeyBytes = Convert.FromBase64String(remotePublicKeyBase64);

            using var remoteKey = ECDiffieHellman.Create();
            remoteKey.ImportSubjectPublicKeyInfo(remoteKeyBytes, out _);

            // Raw ECDH shared secret
            var rawSecret = _localKeyPair.DeriveRawSecretAgreement(remoteKey.PublicKey);

            // SHA-256 hash for uniform key distribution (matches Android side)
            _sharedSecret = SHA256.HashData(rawSecret);

            Debug.WriteLine("SessionCrypto: ECDHE shared secret derived successfully");
        }

        /// <summary>
        /// Encrypt a plaintext string using AES-256-GCM.
        /// Returns Base64 in format: iv_base64:ciphertext_base64
        /// </summary>
        public string Encrypt(string plaintext)
        {
            if (_sharedSecret == null)
                throw new InvalidOperationException("Shared secret not established.");

            var plaintextBytes = Encoding.UTF8.GetBytes(plaintext);
            var iv = RandomNumberGenerator.GetBytes(GcmIvLengthBytes);
            var ciphertext = new byte[plaintextBytes.Length];
            var tag = new byte[GcmTagLengthBytes];

            using var aes = new AesGcm(_sharedSecret, GcmTagLengthBytes);
            aes.Encrypt(iv, plaintextBytes, ciphertext, tag);

            // Combine ciphertext + tag (Android's Cipher.doFinal appends the tag)
            var combined = new byte[ciphertext.Length + tag.Length];
            Buffer.BlockCopy(ciphertext, 0, combined, 0, ciphertext.Length);
            Buffer.BlockCopy(tag, 0, combined, ciphertext.Length, tag.Length);

            var ivBase64 = Convert.ToBase64String(iv);
            var ctBase64 = Convert.ToBase64String(combined);
            return $"{ivBase64}:{ctBase64}";
        }

        /// <summary>
        /// Decrypt a ciphertext string produced by Encrypt (or the Android equivalent).
        /// Input format: iv_base64:ciphertext_base64
        /// </summary>
        public string Decrypt(string encryptedPayload)
        {
            if (_sharedSecret == null)
                throw new InvalidOperationException("Shared secret not established.");

            var parts = encryptedPayload.Split(':', 2);
            if (parts.Length != 2)
                throw new ArgumentException("Invalid encrypted payload format");

            var iv = Convert.FromBase64String(parts[0]);
            var combined = Convert.FromBase64String(parts[1]);

            // Split combined into ciphertext + tag (last 16 bytes = tag)
            var ciphertext = new byte[combined.Length - GcmTagLengthBytes];
            var tag = new byte[GcmTagLengthBytes];
            Buffer.BlockCopy(combined, 0, ciphertext, 0, ciphertext.Length);
            Buffer.BlockCopy(combined, ciphertext.Length, tag, 0, GcmTagLengthBytes);

            var plaintext = new byte[ciphertext.Length];
            using var aes = new AesGcm(_sharedSecret, GcmTagLengthBytes);
            aes.Decrypt(iv, ciphertext, tag, plaintext);

            return Encoding.UTF8.GetString(plaintext);
        }

        /// <summary>
        /// Reset the crypto state (e.g. on disconnect).
        /// </summary>
        public void Reset()
        {
            _localKeyPair?.Dispose();
            _localKeyPair = null;
            _sharedSecret = null;
            Debug.WriteLine("SessionCrypto: Session crypto state reset");
        }
    }
}
